File: demos/go/web/common/kv/kv.go
package kv

import (
	"strconv"
	"time"
)

type KvEngine interface {
	set(key string, val interface{})
	get(key string) interface{}
	remove(key string)
	keys() []string
}

type KvExpiration struct {
	engine         KvEngine
	prefix, suffix string
}

// new expiration
func NewExpiration(engine KvEngine, prefix string) *KvExpiration {
	return &KvExpiration{
		engine: engine,
		prefix: prefix,
		suffix: "-expiration",
	}
}

func NewExpirationWithSuffix(engine KvEngine, prefix, suffix string) *KvExpiration {
	return &KvExpiration{
		engine: engine,
		prefix: prefix,
		suffix: suffix,
	}
}

// set expiration
func (e *KvExpiration) Set(key string, val interface{}, duration int64) {
	e.engine.set(e.genKey(key), val)
	if duration > 0 {
		e.engine.set(e.genExpirationKey(key), time.Now().Unix()+duration)
	}
}

// set expiration next day
func (e *KvExpiration) SetNextDay(key string, val interface{}) {
	now := time.Now()
	nextDay := time.Now().AddDate(0, 0, 1)
	e.Set(key, val, nextDay.Unix()-now.Unix())
}

// set expiration days
func (e *KvExpiration) SetExpirationDays(key string, val interface{}, days int) {
	now := time.Now()
	nextDay := time.Now().AddDate(0, 0, days)
	e.Set(key, val, nextDay.Unix()-now.Unix())
}

// get expiration
func (e *KvExpiration) Get(key string) interface{} {
	if e.isExpired(key) {
		return nil
	}
	return e.engine.get(e.genKey(key))
}

// remove expiration
func (e *KvExpiration) Remove(key string) {
	e.engine.remove(e.genExpirationKey(key))
	e.engine.remove(e.genKey(key))
}

// keys expiration
func (e *KvExpiration) Keys() []string {
	return e.engine.keys()
}

func (e *KvExpiration) genExpirationKey(key string) string {
	return e.prefix + key + e.suffix
}

func (e *KvExpiration) genKey(key string) string {
	return e.prefix + key
}

func (e *KvExpiration) FlushExpired() {
	for _, key := range e.Keys() {
		// check startswith and not endswith
		if key[:len(e.prefix)] == e.prefix && key[len(key)-len(e.suffix):] != e.suffix {
			targetKey := key[len(e.prefix):]
			e.flushExpiredItem(targetKey)
		}
	}
}

// flush expired item
func (e *KvExpiration) flushExpiredItem(key string) bool {
	if e.isExpired(key) {
		e.Remove(key)
		return true
	}
	return false
}

func (e *KvExpiration) isExpired(key string) bool {
	expiration := e.engine.get(e.genExpirationKey(key))
	if expiration == "" {
		return false
	}
	now := time.Now()
	var num int64
	if _, ok := expiration.(int64); ok {
		num = expiration.(int64)
	} else {
		num, _ = strconv.ParseInt(expiration.(string), 10, 64)
	}
	expirationTime := time.Unix(num, 0)

	return now.After(expirationTime)
}

File: demos/py/scraping/requests-demo/kv/kv.py
import time
from datetime import datetime, timedelta
from typing import Any, Dict, List, Optional

class KvEngine:
    def set(self, key: str, val: Any) -> None:
        pass

    def get(self, key: str) -> Any:
        pass

    def remove(self, key: str) -> None:
        pass

    def keys(self) -> List[str]:
        pass

class KvExpiration:
    def __init__(self, engine: KvEngine, prefix: str, suffix: str = "-expiration"):
        self.engine = engine
        self.prefix = prefix
        self.suffix = suffix

    def set(self, key: str, val: Any, duration: int = 0) -> None:
        self.engine.set(self.gen_key(key), val)
        if duration > 0:
            expiration_time = (datetime.now() + timedelta(seconds=duration)).isoformat()
            self.engine.set(self.gen_expiration_key(key), expiration_time)

    def set_with_expiration(self, key: str, val: Any, expiration_time: str) -> None:
        self.engine.set(self.gen_key(key), val)
        self.engine.set(self.gen_expiration_key(key), expiration_time)

    def set_next_day(self, key: str, val: Any) -> None:
        now = datetime.now()
        next_day = now + timedelta(days=1)
        next_day_start = next_day.replace(hour=0, minute=0, second=0, microsecond=0)
        self.set_with_expiration(key, val, next_day_start.isoformat())

    def set_expiration_days(self, key: str, val: Any, days: int) -> None:
        now = datetime.now()
        expiration_time = now + timedelta(days=days)
        self.set_with_expiration(key, val, expiration_time.isoformat())

    def get(self, key: str) -> Optional[Any]:
        if self.is_expired(key):
            return None
        return self.engine.get(self.gen_key(key))

    def remove(self, key: str) -> None:
        self.engine.remove(self.gen_expiration_key(key))
        self.engine.remove(self.gen_key(key))

    def keys(self) -> List[str]:
        return self.engine.keys()

    def gen_expiration_key(self, key: str) -> str:
        return f"{self.prefix}{key}{self.suffix}"

    def gen_key(self, key: str) -> str:
        return f"{self.prefix}{key}"

    def flush_expired(self) -> None:
        for key in self.keys():
            if key.startswith(self.prefix) and not key.endswith(self.suffix):
                target_key = key[len(self.prefix):]
                self.flush_expired_item(target_key)

    def flush_expired_item(self, key: str) -> bool:
        if self.is_expired(key):
            self.remove(key)
            return True
        return False

    def is_expired(self, key: str) -> bool:
        expiration_str = self.engine.get(self.gen_expiration_key(key))
        if expiration_str is None:
            return False
        expiration_time = datetime.fromisoformat(expiration_str)
        now = datetime.now()
        return now > expiration_time
    
    def save(self) -> None:
        if hasattr(self.engine, 'save'):
            self.engine.save()